### PR TITLE
Api 8004 api calls statistics

### PR DESCRIPTION
**Pull request by czeczot**

### CODE REVIEW

**Review by kogierman — CHANGES_REQUESTED** [review]
Generally looks good, I would only like the channel to be passed from outside (similar to i.e. RetryFunc) by SDK consumer, not hardcoded in constructor. This way we are more consistent with "middlewares" convention applied here and also we can avoid needlessly writing to channel when user doesn't care about this data.

**Review by litwinow — COMMENTED** [review]
pls use one-direction channels, also you can send whole structs and not pointers if the structs are lightweight (`APICallStats` is lightweight because it only has `string` [passed as pointer], `bool` and `time.Duration` which is type alias for `int64`)

File: internal/web_api.go [hunk 1]
```diff
@@ -162,6 +180,13 @@ func (a *fileUploadAPI) UploadFile(filename string, file []byte) (string, error)
 		URL string `json:"url"`
 	}
 	err = a.send(req, &resp)
+
+	executionTime := time.Now().Sub(start)
+	if err != nil {
+		a.s.sample(metrics.APICallStats{"upload_file", executionTime, false})
+	} else {
+		a.s.sample(metrics.APICallStats{"upload_file", executionTime, true})
+	}
```
Comment (kogierman) [inline internal/web_api.go:189]: oneline: `a.s.sample(metrics.APICallStats{"upload_file", executionTime, err == nil})`

File: internal/web_api.go [hunk 2]
```diff
@@ -95,8 +98,15 @@ func (a *api) Call(action string, reqPayload interface{}, respPayload interface{
...
 	}
+	err = a.send(req, respPayload)
 
-	return a.send(req, respPayload)
+	executionTime := time.Now().Sub(start)
+	if err != nil {
+		a.s.sample(metrics.APICallStats{action, executionTime, false})
+	} else {
+		a.s.sample(metrics.APICallStats{action, executionTime, true})
+	}
```
Comment (kogierman) [inline internal/web_api.go:108]: oneline

File: internal/web_api.go (hunk 1 above)
Comment (kogierman) [reply internal/web_api.go:189]: #crit check if channel is even set :D

**Review by kogierman — CHANGES_REQUESTED** [review]

File: internal/web_api.go [hunk 3]
```diff
@@ -33,6 +34,7 @@ type api struct {
 	host                 string
 	customHeaders        http.Header
 	retryStrategy        RetryStrategyFunc
+	s                    *stats
```
Comment (litwinow) [inline internal/web_api.go:37]: IMO this `stats` struct is not needed, doesn't provide any value. `sampleChan` could be member of `api` struct and `sample` it's function

File: agent/api.go [hunk 4]
```diff
@@ -16,6 +17,7 @@ type agentAPI interface {
 	SetCustomHost(string)
 	SetCustomHeader(string, string)
 	SetRetryStrategy(i.RetryStrategyFunc)
+	SetSampleChannel(chan<- metrics.APICallStats)
```
Comment (litwinow) [inline agent/api.go:20]: OPT: `SetStatsSink`

File: internal/web_api.go [hunk 5]
```diff
@@ -33,6 +34,7 @@ type api struct {
 	host                 string
 	customHeaders        http.Header
 	retryStrategy        RetryStrategyFunc
+	sampleChan           chan<- metrics.APICallStats
```
Comment (litwinow) [inline internal/web_api.go:37]: OPT: `statsSink`

**Review by litwinow — APPROVED** [review]

**Review by kogierman — APPROVED** [review]

File: agent/api.go [hunk 6]
```diff
@@ -16,6 +17,7 @@ type agentAPI interface {
 	SetCustomHost(string)
 	SetCustomHeader(string, string)
 	SetRetryStrategy(i.RetryStrategyFunc)
+	SetStatsSink(f func(metrics.APICallStats))
```
Comment (litwinow) [inline agent/api.go:20]: Use `StatsSinkFunc` everywhere

File: internal/web_api.go [hunk 7]
```diff
@@ -25,6 +26,12 @@ const apiVersion = "3.2"
 // It returns info whether to retry the request.
 type RetryStrategyFunc func(attempts uint, err error) bool
 
+// StatsSinkFunc if it is set then is invoked by each API request.
+//
+// It accepts one argumet: data - which is metrics data describing
+// requests action type, execution time of request and status of the request
+type StatsSinkFunc func(data metrics.APICallStats)
```
Comment (litwinow) [inline internal/web_api.go:33]: `data` -> `callStats`

File: internal/web_api.go [hunk 8]
```diff
@@ -25,6 +26,12 @@ const apiVersion = "3.2"
 // It returns info whether to retry the request.
 type RetryStrategyFunc func(attempts uint, err error) bool
 
+// StatsSinkFunc if it is set then is invoked by each API request.
+//
+// It accepts one argumet: data - which is metrics data describing
+// requests action type, execution time of request and status of the request
```
Comment (litwinow) [inline internal/web_api.go:32]: don't need to describe what `APICallStats` is here, it should be described where it's defined

File: internal/web_api.go [hunk 9]
```diff
@@ -95,8 +104,14 @@ func (a *api) Call(action string, reqPayload interface{}, respPayload interface{
 		}
 		req.Header.Set(key, val[0])
 	}
+	err = a.send(req, respPayload)
 
-	return a.send(req, respPayload)
+	executionTime := time.Now().Sub(start)
+	if a.statsSink != nil {
+		a.statsSink(metrics.APICallStats{action, executionTime, err == nil})
+	}
```
Comment (litwinow) [inline internal/web_api.go:112]: set `a.StatsSink` to no-op func in the constructor of API and don't check for `nil` here

File: internal/web_api.go [hunk 10]
```diff
@@ -25,6 +26,11 @@ const apiVersion = "3.2"
 // It returns info whether to retry the request.
 type RetryStrategyFunc func(attempts uint, err error) bool
 
+// StatsSinkFunc if it is set then is invoked by each API request.
+//
+// It accepts one argumet: callStats.
```
Comment (litwinow) [inline internal/web_api.go:31]:
```suggestion
// StatsSinkFunc is called after each API method with statistics of that method execution.
```

**Review by litwinow — APPROVED** [review]

**Review by kogierman — APPROVED** [review]

**Pull request merged**